### ISSUE TITLE
Command 'scanamt' not working

### ISSUE THREAD

**Issue by Ilia1** [Ylianst/MeshCentral#516]
MeshCommander finds devices on the network that support AMT
![mesh1](https://user-images.githubusercontent.com/16576224/65878358-5795f600-e3a7-11e9-8f63-59d6c1e9afb9.png)
But command execution 'scanamt' in MeshCentral console does not display any result
![mesh2](https://user-images.githubusercontent.com/16576224/65878365-5c5aaa00-e3a7-11e9-966e-16b56188f71d.png)

**Comment by MailYouLater** [comment]
I can't seem to find any local AMT devices via the scanamt command either, but I can via the "Scan Network" link in an AMT group, I find several.
![image](https://user-images.githubusercontent.com/1279475/65924887-85793b80-e3a3-11e9-87f4-6e9d63321f95.png)
![image](https://user-images.githubusercontent.com/1279475/65924830-5662ca00-e3a3-11e9-803e-c6221bfa5e1a.png)

As a side note, I thought I'd try MeshCommander, but when I tried to install it via NPM I got an old version that doesn't seem to work the way the above screenshot shows.  Also, you may want to fix the binary name in `node_modules\bin`, currently it's `meshcentral`, so if you try to do a global install (`npm install -g meshcommander`), you have to type `meshcentral` to launch MeshCommander, and it'll replace the binary from the `meshcentral` package, so you can only have one work at a time.

**Comment by jsastriawan** [comment]
Scan amt use RMCP ping. This should work on LAN or hybrid mode as long as the subnet is accessible. If you are in WAN mode, it may not work.

**Comment by Ilia1** [comment]
Launched in hybrid mode. Scanamt command does not work. In the group device the network scan finds when specifying the ip address, when specifying the subnet mask does not find.
The meshcentral server is on a different subnet, but an meshagent is installed on that subnet.
I use meshcommander msi from https://www.meshcommander.com/meshcommander

**Comment by jsastriawan** [comment]
The code handling this may be this: https://github.com/Ylianst/MeshCentral/blob/master/meshuser.js#L1969

Will debug later today as soon as I reach office today.

**Comment by MailYouLater** [comment]
I was in LAN only mode (no FQDN defined for WAN or hybrid modes) and my PC, the meshcentral server, and the meshagent are all in the same subnet.
Should I bother to manually set up a meshcentral instance that uses the changes in #518 to test it, or should I just wait till it's merged, and test the next release?

**Comment by jsastriawan** [comment]
Please try #518. I found the issue where scan does not support callback. With this tiny patch, it will work. I have tested it on both Linux and Windows agent. Additionally I fix IP range to remove network and broadcast address the range to avoid irrecoverable error on UDP socket object.
Regards,


Joko

**Comment by jsastriawan** [comment]
In case you want to try this patch. You need to regenerate the minify version of the modified JavaScript. If you can't, you can disable the default use of minified JavaScript using option 'minifycore : 0' in config.json.

Regards,


Joko

**Comment by MailYouLater** [comment]
I downloaded [518.patch](https://patch-diff.githubusercontent.com/raw/Ylianst/MeshCentral/pull/518.patch), patched the file with WinMerge2011, and deleted the `modules_meshcore_min` folder (first I tried just deleting the one `amt-scanner.min.js` from the folder, but then it wouldn't run the `scanamt` command at all) and now it seems to work fine.
![image](https://user-images.githubusercontent.com/1279475/65989063-d50c4580-e43d-11e9-9009-e371da265535.png)

**Comment by jsastriawan** [comment]
Yeah, meshcentral .min.js files are generated using a WebsiteCompiler tool from MeshCommander.  https://www.meshcommander.com/meshcommander

Glad this patch work. Hopefully @Ylianst will merge this soon. I know he is busy now. :)

**Comment by MailYouLater** [comment]
Agreed.
Also ... this is probably a question for @Ylianst, but ...  [MeshCentral](/Ylianst/MeshCentral) and [MeshAgent](/Ylianst/MeshAgent) each have a GitHub Repo.  Why doesn't MeshCommander?

**Comment by Ylianst** [comment]
@MailYouLater Oh dear! I just published [MeshCommander v0.7.9-a on NPM](https://www.npmjs.com/package/meshcommander) with a fix for the binary name, it was incorrectly "MeshCentral" now corrected to "MeshCommander". I also updated the latest MeshCommander code.

Your right that I never added MeshCommander to GitHub and I really should at some point. It's a bit difficult when I already have so many issues on MeshCentral, but I will do that at some point. I will need to also put in GitHub the WebSiteCompiler that processes MeshCommander code. All this is a bunch of work.

Looking at @jsastriawan patch now!

**Comment by MailYouLater** [comment]
Thanks for updating MeshCommander on NPM. I now have 0.7.9-a, though the interface still looks different, and I don't see how to scan for AMT devices with it. I'm guessing there's some UI elements, and/or features that are exclusive to the version installed by the MSI file available from MeshCommander.com.

Regarding getting a MeshCommander and/or WebSiteCompiler repo on GitHub.  No worries, I'm not in a big rush for either. Thanks for all the work you do, and I'll look forward to seeing it/them when you have time to put it/them up on GitHub.

**Comment by Ylianst** [comment]
First, MeshCentral v0.4.1-p has the patch from @jsastriawan and the "scanamt" command will not works. I also added "amtscan" to MeshCmd. Hopefully that will be useful:

![MeshCmd-AmtScan](https://user-images.githubusercontent.com/1319013/66064206-d26f2600-e4f9-11e9-97b4-4354af9be1c5.png)

For MeshCommander, your right that the NPM version does not have the scan feature, the NW.js (MSI) does. The NW.js version can directly send UDP packets and perform the scan, the NPM version runs in a browser and the browser can't do that. I would need to have the mini-server perform the scan and send the results back to MeshCommander in the browser, it's just more work I would need to do sometime.

**Comment by MailYouLater** [comment]
I removed my `node_modules` folder, reinstalled MeshCentral from NPM, and was able to use the `scanamt` command in the agent console to find AMT capable machines on the network. 🎉
I haven't tested the `amtscan` command in MeshCMD, but shouldn't both of these have the same name? Either `scanamt` or `amtscan` for both of these, instead of one tool using one and the other using the other?

Also, I did notice that if I missed the `0` before the subnet mask (i.e. `192.168.2/24` instead of `192.168.2.0/24`), it acted like it was working, but it couldn't find any AMT devices. This was my mistake of course, but I don't know if it was scanning the wrong subnet, or just completely not working. Perhaps it would be a good idea to have it print out what IP range it's scanning, so the user can see what's happening if it's not doing what they intended.
